### PR TITLE
Implement multiplicative method for total loss computation and new `alloc_rule=3`

### LINKED ISSUE #1099: Implement multiplicative method for total loss computation
This issue is to track progress in implementing a new gulcalc allocrule for total peril loss using the multiplicative method in gulpy https://github.com/OasisLMF/ktools/issues/118.

### PR DESCRIPTION
**Pull request by mtazzari**
This PR implements a new way of computing the total losses using multiplicative expression for sub-peril losses and, respectively, a new alloc_rule value: 3.

This PR tracks implementation of this feature in gulpy. Details are in https://github.com/OasisLMF/ktools/issues/118

Fix #1099 

TODO:

- [x] check that gulpy produces same results as before for alloc_rule = 0, 1, 2.

### Release notes feature title
This PR implements a new formula for computing the total losses using multiplicative expression for sub-peril losses.
This implementation is accessible using the new alloc_rule = 3, i.e. with:
```bash
gulpy -a3 ...
```
After computing total losses using the new formula, back-allocation in proportion to the losses is applied to the sub-peril losses, similarly to the back-allocation applied with `alloc_rule=1`.

More details are in https://github.com/OasisLMF/ktools/issues/118

### CODE REVIEW

**Comment by mtazzari** [comment]
@sstruzik FYI, in manager.py I also fix a typo that we left when we refactored gulpy the first time. We changed the `losses` array to store the negative sidx samples at the end of the array, instead of at the beginning using my padding, and we forgot that split_tiv only had to be applied to the positive sidx:
```py
for sample_i in range(1, losses.shape[0]):
```
should have been
```py
for sample_i in range(1, losses.shape[0] - 5):
```

**Comment by johcarter** [comment]
is split_tiv anything to do with the sub-peril backallocation, @mtazzari ? from memory, I think the alloc rule should also apply to sidx -1, -3, -5

File: oasislmf/pytools/gul/core.py [hunk 1]
```diff
@@ -116,6 +116,33 @@ def split_tiv(gulitems, tiv):
...
+    total loss as tiv * (1 - (1-A)*(1-B)*(1-C)...), where A, B, C are damage ratios
+    computed as the ratio between a sub-peril loss and the tiv. Sub-peril losses
+    in gulitems are always back-allocated proportionally to the losses.
+
+    Args:
+        gulitems (numpy.array[oasis_float]): array containing losses of all items.
+        tiv (oasis_float): total insured value.
+    """
+    Ngulitems = gulitems.shape[0]
+    total_loss = 1.
```
Comment (sstruzik) [inline oasislmf/pytools/gul/core.py:131]: I would call that insure_value_left or undamaged_value or 1_minus_total_loss.
I prefer to avoid confusion when reading as until line 135 it is not the total loss

**Comment by mtazzari** [comment]
@johcarter 
> is split_tiv anything to do with the sub-peril backallocation, @mtazzari ? from memory, I think the alloc rule should also apply to sidx -1, -3, -5

you're right, we had an issue open in ktools and oasislmf as well: https://github.com/OasisLMF/OasisLMF/issues/1003
Current implementation in gulcalc applies split_tiv to all sidx samples.
I suggest we keep it like this in this PR and we address this in general in another PR.

File: oasislmf/pytools/gul/core.py [hunk 2]
```diff
@@ -116,6 +116,33 @@ def split_tiv(gulitems, tiv):
...
+    for i in range(Ngulitems):
+        total_loss *= 1. - gulitems[i] / tiv
+
+    total_loss = 1. - total_loss
+    total_loss *= tiv
+    if total_loss > 0.:
```
Comment (sstruzik) [inline oasislmf/pytools/gul/core.py:137]: I think this is the same and a bit less computation:
if insure_value_left < 1:
          f = 1/(1-insure_value_left)
          for j in range(Ngulitems):
            # editing in-place the np array
            gulitems[j] *= f
Can you double check?

File: oasislmf/pytools/gul/manager.py [hunk 3]
```diff
@@ -379,17 +379,23 @@ def write_losses(event_id, sample_size, loss_threshold, losses, item_ids, alloc_
...
     """
-    # note that Nsamples = sample_size + NUM_IDX
-
     if alloc_rule == 2:
         losses[1:] = setmaxloss(losses[1:])
 
     # split tiv has to be executed after setmaxloss, if alloc_rule==2.
-    if alloc_rule >= 1 and tiv > 0:
+    if tiv > 0:
         # check whether the sum of losses-per-sample exceeds TIV
```
Comment (sstruzik) [inline oasislmf/pytools/gul/manager.py:387]: remove comments as not relevant anymore

Comment (mtazzari) [reply oasislmf/pytools/gul/manager.py:387]: ok done

File: oasislmf/pytools/gul/core.py (hunk 2 above)
Comment (mtazzari) [reply oasislmf/pytools/gul/core.py]: doesn't apply anymore I think

File: oasislmf/pytools/gul/core.py (hunk 1 above)
Comment (mtazzari) [reply oasislmf/pytools/gul/core.py]: changed `total_loss` to `undamaged_value`

**Review by sstruzik — APPROVED** [review]
looks good
